### PR TITLE
MGDAPI-2906 - add byovpc creation logic to cluster/create

### PR DESCRIPTION
**Pull request by laurafitzgerald**
WHAT
add steps/script updates to allow for the creation of byovpc with one private and one public subnet.

VERIFICATION
follow the steps to create a vpc in AWS and create the cluster. 
No errors should occur and you should end up with a byovpc cluster.

cluster.json should have the following config present after generation
```
....
  "nodes": {
.....
    "availability_zones": [
      "us-east-1d"
    ]
......
"aws": {
......
  "subnet_ids": [
      "subnet-00fdce7e7bb7498af",
      "subnet-03dfd2dc092d1851c"
    ],
    "private_link": false
....
```

Verify that the generated cluster.json can be used via ocm cli to create an OSD.
This can be tested using `make ocm/cluster/create`

### CODE REVIEW

File: docs/ocm/README.md [hunk 1]
```diff
@@ -44,6 +44,25 @@ If you want your BYOC cluster to span multiple availability zones then set the `
...
+
+If you do not already have a VPC created go to aws and create a VPC using the VPC Wizard available from the VPC Dashboard.
+
+Create a VPC with Public and Private Subnets. Note the CIDR Range ip as you will need this for your machine CIDR, or alternatively use 10.11.0.0/16 which will satisfy the default machine_cidr
+
+Provide a name, chose an availability zone and update the subnets to suitable values. Chose an elastic ip allocation and chose CreateVPC
+
+Export the following values
+```
+export BYPVPC=true
```
Comment (austincunningham) [inline docs/ocm/README.md:59]: typo here I think

**Comment by austincunningham** [comment]
## Verified
setup a vpc via wizard
Screenshot from 2021-12-17 12-16-50

```bash
export BYOVPC=true
export PRIVATE_SUBNET_ID=10.11.1.0/24
export PUBLIC_SUBNET_ID=10.11.0.0/24
export AVAILABILITY_ZONES=us-east-2a
export OCM_CLUSTER_REGION=us-east-2a
export AWS_ACCOUNT_ID=<redacted>        
export AWS_ACCESS_KEY_ID=<redacted> 
export AWS_SECRET_ACCESS_KEY=<redacted> 
make ocm/cluster.json
```
```json
{
  "display_name": "rhmi-211217-1303",
  "ccs": {
    "enabled": true
  },
  "managed": true,
  "multi_az": false,
  "name": "rhmi-211217565",
  "network": {
    "machine_cidr": "10.11.128.0/23",
    "service_cidr": "10.11.0.0/18",
    "pod_cidr": "10.11.64.0/18",
    "host_prefix": 23
  },
  "nodes": {
    "compute": 8,
    "compute_machine_type": {
      "id": "m5.xlarge"
    },
    "availability_zones": [
      "us-east-2a"
    ]
  },
  "region": {
    "id": "us-east-2a"
  },
  "expiration_timestamp": "2021-12-17T17:03:02Z",
  "api": {
    "listening": "external"
  },
  "aws": {
    "access_key_id": "<redacted> ",
    "secret_access_key": "<redacted> ",
    "account_id": "<redacted> ",
    "subnet_ids": [
      "10.11.1.0/24",
      "10.11.0.0/24"
    ],
    "private_link": false
  }
}
```

File: docs/ocm/README.md (hunk 1 above)
Comment (austincunningham) [reply docs/ocm/README.md:59]:
```suggestion
export BYOVPC=true
```

**Comment by laurafitzgerald** [comment]
@austincunningham did the above work for you? I had to supply the subnet ids rather than the cidr blocks to aws.subnetids

**Comment by austincunningham** [comment]
@laurafitzgerald didn't test it beyond it updated the cluster.json with values in the env vars.

**Comment by laurafitzgerald** [comment]
ah okay, I think the verification should alos include that the cluster.json can be successfully used to create the OSD.

**Comment by austincunningham** [comment]
attempted to verify but there seems to be an issue identifying the public subnet
```bash
make ocm/cluster/create
Sending a request to OCM to create an OSD cluster
{
  "kind": "Error",
  "id": "400",
  "href": "/api/clusters_mgmt/v1/errors/400",
  "code": "CLUSTERS-MGMT-400",
  "reason": "The following errors occurred in the subnets provided: Expected a public subnet to be provided in each zone, instead all subnets specified in zones [us-east-2a] are private.",
  "operation_id": "1fe964eb-6716-4afd-9fec-a7030d8a85de"
}
make: *** [make/ocm.mk:37: ocm/cluster/create] Error 1```

**Comment by laurafitzgerald** [comment]
Confirmed the above is caused by a bug in ocm. They are working to fix it so putting the verification of this change on hold for now.

**Comment by laurafitzgerald** [comment]
@austincunningham do you have time to see if the issue is resolved? or we could request an intern to take a look. maybe @Patryk-Stefanski might be interested as he's looking at a similar ticket.

**Comment by Patryk-Stefanski** [comment]
Ran these commands to get the cluster.json file.
```
export BYOVPC=true
export BYOC=true
export PRIVATE_SUBNET_ID=subnet-00fdce7e7bb7498af
export PUBLIC_SUBNET_ID=subnet-03dfd2dc092d1851c
export AVAILABILITY_ZONES=us-east-1d
export OCM_CLUSTER_REGION=us-east-1
export AWS_ACCOUNT_ID=  *************     
export AWS_ACCESS_KEY_ID= **************
export AWS_SECRET_ACCESS_KEY= ****************
make ocm/cluster.json
```

cluster.json file contents 
```
{
  "display_name": "rhmi-220119-1101",
  "ccs": {
    "enabled": true
  },
  "managed": true,
  "multi_az": false,
  "name": "rhmi-2201192290",
  "network": {
    "machine_cidr": "10.0.0.0/23",
    "service_cidr": "10.11.0.0/18",
    "pod_cidr": "10.11.64.0/18",
    "host_prefix": 23
  },
  "nodes": {
    "compute": 8,
    "compute_machine_type": {
      "id": "m5.xlarge"
    },
    "availability_zones": [
      "us-east-1d"
    ]
  },
  "region": {
    "id": "us-east-1"
  },
  "expiration_timestamp": "2022-01-19T15:01:56Z",
  "api": {
    "listening": "external"
  },
  "aws": {
    "access_key_id": "****************",
    "secret_access_key": "**********************",
    "account_id": "*******************",
    "subnet_ids": [
      "subnet-00fdce7e7bb7498af",
      "subnet-03dfd2dc092d1851c"
    ],
    "private_link": false
  }
}
```
`Note: The machine CIDR  in  the cluster.json  file was changed manually to avoid overlap issues`

Then I ran `make ocm/cluster/create` and can verify that the cluster was installed successfully.

image

When reading through the documentation for creating BYOVPC think it's a bit confusing. could we make it more clear that you need to set up BYOC in order to set up BYOVPC?

**Comment by laurafitzgerald** [comment]
thanks for the feedback @Patryk-Stefanski one slightly unrelated question is, are we defaulting to 8 nodes. I think we can reduce that to 5. @psturc wdyt?

**Comment by psturc** [comment]
@laurafitzgerald changing the default number of nodes would have an impact on our pipelines, because some of them are relying on the default value. If you think 8 is too much and it's worth to change it, I think it's worth to create a separate ticket for this change.

**Comment by laurafitzgerald** [comment]
@Patryk-Stefanski I added a note to address your feedbakc. If you are happy with that can you drop `/lgtm` into a comment. Or if there is any other feedback, let me know. 👍

**Comment by Patryk-Stefanski** [comment]
Looks  good .
/lgtm

**Pull request merged**